### PR TITLE
[7.4-only] LPS-121960 Migrate v2 usages of clay:vertical-card in depot/depot-web

### PR DESCRIPTION
**Pull request by julien**
**IMPORTANT**: This pr needs to be sent to @liferay-lima after internal review

In this change, we replace `<clay:vertical-card-v2 />`(clay 2.x: Metal+soy) with
`<clay:vertical-card />` (clay 3.x: React)

**Test plan**:

- Fetch this branch
- Deploy the `depot-web` and `frontend-taglib-clay` modules
- From the global menu, in the **Applications** tab, select **Content** > **Asset Libraries**
- Click on the **+** button in the top right corner of the page to
create a new asset library
- Enter a name for this asset library
- Scroll down to the end of the page and click on the **Save** button
- Assert that the asset library has been created and that a card is
visible and working as expected.

### CODE REVIEW

**Comment by julien** [comment]
ci:test:sf

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js
Comment (julien) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js:18]: This is the warning I'm talking about in case it's not clear

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld
Comment (julien) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld:4373]: @carloslancha, @markocikos I wasn't sure this was needed but since [other](https://github.com/liferay/liferay-portal/blob/2ba32f018d39b3eaacd32ff6c0fed7cfb806cff7/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld#L307-L312) [tags](https://github.com/liferay/liferay-portal/blob/2ba32f018d39b3eaacd32ff6c0fed7cfb806cff7/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld#L1042-L1047) have it I though it wouldn't hurt.

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld:4373]: Maybe it also needs 

```xml
<attribute>
    <description></description>
    <name>propsTransformerServletContext</name>
    <required>false</required>
    <rtexprvalue>true</rtexprvalue>
</attribute>
```

but I'll wait for your comments

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js
Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js:18]: And by the way, I have a feeling that this error





 that we've been seeing for a while could be fixed if [`DropdownMenu.js`](https://github.com/liferay/liferay-portal/blob/b50d81b60bf1437cc1418227d8d96eaefefce314/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js) would use the [`getDataAttributes`](https://github.com/liferay/liferay-portal/blob/b50d81b60bf1437cc1418227d8d96eaefefce314/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js) function instead of it's own - but that doesn't seem to be part of the scope of this PR.
If you agree though, I can go ahead and create a ticket and send a fix for that.

File: modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java
Comment (markocikos) [inline modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java:82]: Besides here, `DepotAdminWebKeys.DEPOT_ENTRY_DROPDOWN_DEFAULT_EVENT_HANDLER` is used only in two `<clay:dropdown-actions>`.  Looking [here](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld#L863), `<clay:dropdown-actions>` also has `propsTransformer` attribute. If we can replace those usages, we can clean up everything related to default event handler, like delete `DepotEntryDropdownDefaultEventHandler.es.js` file and the Java constant.

File: modules/apps/depot/depot-web/src/main/resources/META-INF/resources/js/DepotEntryVerticalCardPropsTransformer.js
Comment (markocikos) [inline modules/apps/depot/depot-web/src/main/resources/META-INF/resources/js/DepotEntryVerticalCardPropsTransformer.js:25]: We can use `openModal` here, imported from `frontend-js-web`.
```
openModal({
    title: Liferay.Language.get('permissions'),
    url: itemData.permissionsDepotEntryURL,
})
```

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld
Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld:4373]: Agree on adding `propsTransformer`.

Why would we add `propsTransformerServletContext`? Are we using it somewhere but it falls under dynamic attributes, or is it commonly used in other tags that also have `propsTransformer`?

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js
Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js:18]: > If you agree though, I can go ahead and create a ticket and send a fix for that.

:+1:

**Review by markocikos — COMMENTED** [review]
Few comment inline. This  `propsTransformer` business looks to be the biggest pain point in usage updates. But, it does make sense to do it.

File: modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java
Comment (julien) [reply modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java:82]: Sounds like a good idea.

File: modules/apps/depot/depot-web/src/main/resources/META-INF/resources/js/DepotEntryVerticalCardPropsTransformer.js
Comment (julien) [reply modules/apps/depot/depot-web/src/main/resources/META-INF/resources/js/DepotEntryVerticalCardPropsTransformer.js:25]: I knew there was something we could import I just couldn't remember the name :+1:  :rofl:

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld
Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld:4373]: Yes, other tags that have the `propsTransformer` attribute usually have `propsTransformerServletContext`, so I just wanted to make that identical.

File: modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java
Comment (julien) [reply modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java:82]: One thing I don't really like is that  [`DepotEntryDropdownDefaultEventHandler`](https://github.com/liferay/liferay-portal/blob/826c5123ca36f821eba3406147aff60ff3de9ef0/modules/apps/depot/depot-web/src/main/resources/META-INF/resources/js/DepotEntryDropdownDefaultEventHandler.es.js), and the "new" [`DepotEntryVerticalCardPropsTransformer`](https://github.com/liferay-frontend/liferay-portal/blob/0a1799a69a67b4381c2c0009aa8dbe50407a8a52/modules/apps/depot/depot-web/src/main/resources/META-INF/resources/js/DepotEntryVerticalCardPropsTransformer.js) are identical, *but* they are used by two components that have the "same" property (i.e. that prop has the same purpose) but with a different name:
[`DropDownWithItems`](https://github.com/liferay/clay/blob/e0a84fb91adebbf8005fce3bcb15abdb24ebd12e/packages/clay-drop-down/src/DropDownWithItems.tsx#L296) uses `items`, but [`CardWithInfo`](https://github.com/liferay/clay/blob/master/packages/clay-card/src/CardWithInfo.tsx#L108) uses `actions`.

Of course, we can do something like this:

```js
import {openModal} from 'frontend-js-web';
import confirmDepotEntryDeletion from './confirmDepotEntryDeletion.es';

const ACTIONS = {
	deleteDepotEntry(itemData) {
		if (confirmDepotEntryDeletion()) {
			submitForm(document.hrefFm, itemData.deleteDepotEntryURL);
		}
	},

	permissionsDepotEntry(itemData) {
		openModal({
			title: Liferay.Language.get('permissions'),
			url: itemData.permissionsDepotEntryURL,
		});
	},
};

function mapActionsOrItems(item) {
	return {
		...item,
		onClick(event) {
			const action = item.data?.action;

			if (action) {
				event.preventDefault();

				ACTIONS[action](item.data);
			}
		},
	};
}

export default function propsTransformer({actions, items, ...otherProps}) {
	const prop = actions ? 'actions' : items ? 'items' : null;

	return {
		...otherProps,
		[prop]: actions ? actions.map(mapActionsOrItems) : items.map(mapActionsOrItems),
	};
}
```

But it feels like a hack and fragile.

Do you guys have any idea how to improve that? (/cc @wincent ) 

(Opening an issue in clay to rename the prop in one of the two components, sending a PR, waiting for a release and update in DXP and updating the usages doesn't seen like an option)

Comment (julien) [reply modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java:82]: I guess we could do something like this too

```js
export default function propsTransformer({actions, items, ...otherProps}) {
	const result = {
		...otherProps,
	};

	if (actions) {
		result.actions = actions.map(mapActionsOrItems);
	}
	else if (items) {
		result.items = items.map(mapActionsOrItems);
	}

	return result;
}
```

Comment (julien) [reply modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java:82]: So that's what I chose to go with in [
171f3f5](https://github.com/liferay-frontend/liferay-portal/pull/495/commits/171f3f5304fdc7f31f90bb37c09aec30661d0310)

Comment (wincent) [reply modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java:82]: If we're golfing here, technically speaking you don't need to make a copy of `otherProps` and name it `result`, because `otherProps` is _already_ a new object that you own and are free to mutate.

```js
export default function propsTransformer({actions, items, ...props}) {
    if (actions) {
        props.actions = actions.map(map);
    }
    else if (items) {
        props.items = items.map(map);
    }

    return props;
}
```

I don't know if you want to check against being improperly called (ie. with `actions` _and_ `items`) by throwing or logging an error. 🤷

Comment (julien) [reply modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java:82]: @wincent thanks, that's nicer.

> I don't know if you want to check against being improperly called (ie. with actions and items) by throwing or logging an error. :shrug:

sounds like a good idea

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld
Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld:4373]: No idea what is `propsTransformerServletContext` for. I'd say no to add stuff until we need it, but in this case I think we need some context. @jbalsas please?

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js
Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js:18]: As we talked offline I'd not go with this approach since we're renaming stuff and that could cause confusion in developers. For example:

In Java they add to the items some data like `action` or `deleteDepotEntryURL`.

That that is mean to be used in the `propsTransformer`

```
submitForm(document.hrefFm, itemData.deleteDepotEntryURL);
```
If we transform those data attributes to lowerCase then we will receive `itemData.deletedepotentryurl` there, which could be confusion for the developer.

Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js:18]: What I don't know is how is this working right now 😅

File: modules/apps/depot/depot-web/src/main/resources/META-INF/resources/js/DepotEntryPropsTransformer.js
Comment (carloslancha) [reply modules/apps/depot/depot-web/src/main/resources/META-INF/resources/js/DepotEntryPropsTransformer.js:57]: Ok, I just saw it in [this thread](https://github.com/liferay-frontend/liferay-portal/pull/495#discussion_r521463064), it was resolved so I didn't see it.

I'm starting to hate github...

cc @wincent

Comment (julien) [reply modules/apps/depot/depot-web/src/main/resources/META-INF/resources/js/DepotEntryPropsTransformer.js:57]: @carloslancha see [this comment](https://github.com/liferay-frontend/liferay-portal/pull/495#discussion_r521960413)

Comment (wincent) [reply modules/apps/depot/depot-web/src/main/resources/META-INF/resources/js/DepotEntryPropsTransformer.js:57]: > I'm starting to hate github...

You know what I say to that...

let-the-hate-flow-through-you

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js
Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js:18]: @carloslancha I agree that renaming stuff is probably a dangerous move.

if we're accessing an element's attributes (I don't know if that's the case though) there's no difference, i.e.

```js
var el = document.createElement('div');
el.setAttribute('data-myStuff', 'foo');
console.log(el.getAttribute('data-myStuff')); // "foo" 
console.log(el.getAttribute('data-mystuff')); // "foo" 
```

That said, and as talked on slack,  if we changed the case of the attributes from Java, we'd still be renaming those attributes right?

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld
Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld:4373]: @carloslancha I know what it's for :smile:

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld:4373]: @carloslancha see [here](https://github.com/liferay/liferay-portal/blob/9d49616b01cc7f892e544d79db90fcc42775c08b/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/internal/servlet/taglib/BaseContainerTag.java#L319-L323):

What I understand is that if a `propsTransformerServletContext` is passed, it will be used to resolve (load) the `propsTransformer` js module.

>  I'd say no to add stuff until we need it, 

Since `ButtonTag` and `DropdownMenuTag` have both `propsTransformer` and `propsTransformerServletContext` as attributes, I'm just following the same convention and for me it makes sense

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js
Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js]: Ok, so after having analysed this and discussed it with @carloslancha we decided we're not going to change anything in the `getDataAttributes` function because we shouldn't, at the risk of breaking things.

We're also thinking that we are using the `data` attributes incorrectly so we could fix that in the future.

**Comment by julien** [comment]
once everything's OK, I'll squash these commits before sending to @liferay-lima

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js
Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js]: To expand what @julien said

The thing is we were adding some `data` to the items in order to use that that in the `defaultEventHandlers` (like the action name and needed parameters like `deleteDepotEntryURL`), but they're not meant to be rendered as data-attributes, and we don't needed now that we have the `propsTransformer` mechanism.

For now I'll leave it as it is, it's just a warning, but I'd suggest to create a new pattern for this. For example, we can add all this info to an object called `additionalData`, that we will use in the `propsTransformer` but that we will not pass through in the js component (`VerticalCard.js` in this case).

In this way we can continue using `data` for passing `data-attributes` when needed and `additionalData` only for those we will need to use in a `propsTransformer`.

Makes sense?

Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js]: cc @jbalsas

Comment (jbalsas) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js]: Not sure where to answer... this is quite a long thread and I haven't read it through.

https://html.spec.whatwg.org/#custom-data-attribute

> A custom data attribute is an attribute in no namespace whose name starts with the string "data-", has at least one character after the hyphen, is XML-compatible, and contains no ASCII upper alphas.

Which is what @julien was exposing in https://github.com/liferay-frontend/liferay-portal/pull/495#discussion_r522088968

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js]: > Not sure where to answer... 

Yes you're right the "threads" in this PR are really hard to see so from now on @carloslancha and I decided we shouldn't make the conversations as resolved because it makes it hard to see what's going on.

Thanks for the answer, I should have checked the spec and posted that too.
I personally think it would be "nicer" if we didn't have those react warning because seeing that output in the devtools is not a great experience.

@carloslancha do you agree if we change those attributes to lowercase as I was doing initially [here](https://github.com/liferay-frontend/liferay-portal/blob/0a1799a69a67b4381c2c0009aa8dbe50407a8a52/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js#L18)?

Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js]: As we talked offline, I'd leave it as it is, is a warning about something that is not completely right and that will be solved when this gets migrated to use `propsTransformer` mechanism.

File: modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java
Comment (matuzalemsteles) [reply modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java:82]: @julien I ended up being late for this, I think I missed the Github notification 🤷‍♂️.

I think they are different contexts, keeping the same name of `items` for `ClayCardWithInfo` may not mean much, but` actions` adds more context to what to expect from this API, in my view and I think that also when we developed this, `items` in `ClayDropDownWithItems` makes sense because it is the main API for the component. So I'm not sure if it would be a good idea to keep the same API name on both components. Does that make sense to you?

Just remembering that we cannot change the name of the APIs now, we would have to maintain compatibility in any way or remove it only in v4 if we reach an agreement.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js
Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js]: @carloslancha yes agreed. I [reverted that change](https://github.com/liferay-frontend/liferay-portal/pull/495/commits/9ce43829036624b42a05986dbb270d7cecd86629) after our conversation.

File: modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java
Comment (julien) [reply modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/servlet/taglib/clay/DepotEntryVerticalCard.java:82]: @matuzalemsteles no problem we can keep the name `actions` for Cards and `items` for dropdowns.
 
> Just remembering that we cannot change the name of the APIs now,

Yes I know, I think that's also one thing we shouldn't have done between clay 2.x and clay 3.x (on the JS side) to make the replacements easier, but we can deal with it.

**Comment by julien** [comment]
Rebased with `master`

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js
Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/get_data_attributes.js]: > What I don't know is how is this working right now

It works because **most** browsers do `toLowerCase` automatically.

**Review by markocikos — APPROVED** [review]
LGTM!

I do think that we should lowercase `data` attributes to avoid adding invalid HTML, but that can be done in a separate dedicated task. I bet we are doing this in many other places in portal.

**Comment by julien** [comment]
> I do think that we should lowercase data attributes to avoid adding invalid HTML, but that can be done in a separate dedicated task. I bet we are doing this in many other places in portal.

That's a good point @markocikos.

What we think with @carloslancha is that we shouldn't do that for now.

We are doing that in other places in DXP like you mentioned, and the reason is that before with clay 2.x we would use `data-*` attributes to store additional data on the component, with clay 3.x we can now use `additionalProps`, the only thing that we'd have to think about is how to filter these props, because they are usually passed to Clay components and rendered on elements that don't necessarily need them.

**Comment by julien** [comment]
@carloslancha just so you know, even though this is ready to review, I'm not running the ci:test:relevant test because I'm waiting for #507 to be merged, it could affect the tests. Should I convert back to draft?

**Comment by carloslancha** [comment]
Let's run the tests and see if they complain. If not we can forward it.

**Comment by carloslancha** [comment]
ci:test:sf

**Comment by carloslancha** [comment]
ci:test:relevant

**Comment by julien** [comment]
> `[exec] Execution failed for task ':dxp:apps:commerce:commerce-dashboard-web:compileJava'.`

Doesn't seem related

**Comment by julien** [comment]
ci:test:relevant

**Comment by carloslancha** [comment]
@julien this needs to be forwared to Bchan or to the team that owns the app?

**Comment by julien** [comment]
@carloslancha this needs to be forwarded to @liferay-lima, but don't worry I can take care of it, thanks for asking

**Comment by julien** [comment]
Depends on #522 - putting back in draft mode

**Comment by carloslancha** [comment]
We don't need to transform `DefautlEventHandler` to `PropsTransformer` because the `actions` != `items` stuff. It will work once #522 is merged.

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

**Comment by julien** [comment]
depends #534

**Comment by julien** [comment]
re converted to draft to avoid merging

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

**Comment by julien** [comment]
ci:stop

**Comment by julien** [comment]
ci:close

**Comment by julien** [comment]
Sent to https://github.com/liferay-lima/liferay-portal/pull/1102

**Pull request closed without merging**
